docs/styleguides/python.md

```diff
@@ -0,0 +1,231 @@
+
+##### Python Version Requirement
+
+This style guide is written for Python3.5 and above. As new features of the language get standardized, this style guide
+can be revised to adopt new features.
+
+#### Base Style Guide
+
+* [Google Python Style Guide Revision 2.59]
+
+
+#### DC/OS Specific extensions to Base Style Guide
+
+* Import Requirements
+    * Adopt the base style guide preferences.
+
+        ```
+        Use import x for importing packages and modules.
+        Use from x import y where x is the package prefix and y is the module name with no prefix.
+        Use from x import y as z if two modules named y are to be imported or if y is an inconveniently long name.
+        ```
+    * Use `__init__.py` only to represent python packages.
+        * Prefer empty `__init__.py` modules. Keep definitions in submodules.
+
+    _Rationale_: Keeping `__init__.py` empty helps in avoid unnecessary imports when you import a submodule from
+    package. Plus,it helps in navigating the code base you know that an object is strictly defined in a submodule and
+    not aliased in `__init__.py`
+
+* Import formatting
+
+    * The types of imports should be separated by single line and order should be.
+
+        1. standard library imports.
+        2. third-party imports.
+        3. application specific imports.
+
+    * The `from` style import and `import` statements should be separated by single line too.
+    * Each group of these imports should be sorted.
+
+    _Rationale_:  Readablity and code-foldability are reasons for this separation.
+
+For Example
+
+```
+import json
+import uuid
+
+from subprocess import CalledProcessError
+from typing import Callable, List
+
+import retrying
+
+from retrying import retry, RetryError
+from teamcity.messages import TeamcityServiceMessages
+
+import test_util.aws
+import test_util.cluster
+
+from pkgpanda.util import logger, write_string
+```
+
+
+* [Packages](https://google.github.io/styleguide/pyguide.html?showone=Packages#Packages)
+    * Import each module using the full pathname location of the module.
+    * No Relative Imports.
+
+  _Rationale_ - Relative imports are hard to follow and hard to write in consistent manner. And if you are using IDEs
+  to navigate the code base, it is much easier to get your code navigation right with absolute imports from the source
+  root.
+
+* Maximum line length is 120 characters.
+    * Following are the exceptions to this rule. These types can be of arbitrary length.
+        * Long import statements.
+        * URLs in comments.
+        * Any sequence of chars, like embedded tokens which would better be in a single line.
+
+   _Rationale_: At the time this style guide was written, the existing code base followed 120 column width guideline.
+   We are keeping up with that.
+
+* Indentation
+
+    * Use 4-space indentation.
+    * For long list of arguments, when the number of arguments exceed the column limit 120 chars, use double-indented
```

Comment:
> In any case where the hanging indent matches the next logical indent, you need to double indent.

The statement in the style guide is: For long list of arguments, use double-indented hanging-indentation.

i.e, irrespective of any situation, if you are using hanging indent, use double indent. The rationale is that, it gives a visual significance to indentation and indicates that it is used for wrapping and not as a context block.

Here is a comparison for clarity

* exceeds 120 chars # BAD.

```
def method():
    
    installer.setup_remote(tunnel=bootstrap_host_tunnel, installer_path=cluster.ssher.home_dir + '/dcos_generate_config.sh', download_url=installer_url)

```
* single indented - Not recommended. # BAD.
Note hanging indentation aligns with the logical indentation of the program.

```
def method():
    if installer.prereqs.ok:
        print("pre-reqs satisfied")

    installer.setup_remote(
        tunnel=bootstrap_host_tunnel,
        installer_path=cluster.ssher.home_dir + '/dcos_generate_config.sh',
        download_url=installer_url)

    if installer.setup.ok:
        print("remote setup successful.")
    else:
        print("remote setup successful.")
```
 
* double indented - Recommended # GOOD. 
   Note the visual clarity.



```
def method():
    if installer.prereqs.ok:
        print("pre-reqs satisfied")

    installer.setup_remote(
            tunnel=bootstrap_host_tunnel,
            installer_path=cluster.ssher.home_dir + '/dcos_generate_config.sh',
            download_url=installer_url)

    if installer.setup.ok:
        print("remote setup successful.")
    else:
        print("remote setup successful.")
```

* Even if the `installer.setup_remote` is the last line of the program, I think, it is better served with double indented hanging indent.

If the statement of the rule is not-clean / ambiguous, please suggest a improvement. I can also consider adding this example to the style guide itself.